taichi/codegen/llvm/codegen_llvm.cpp

```diff
@@ -2757,6 +2749,49 @@ llvm::Type *TaskCodeGenLLVM::get_real_func_ret_type(Function *real_func) {
   return llvm::StructType::get(*llvm_context, tps);
 }
 
+void TaskCodeGenLLVM::create_return(llvm::Value *buffer,
+                                    llvm::Type *buffer_type,
+                                    const std::vector<Stmt *> &elements,
+                                    const Type *now_type,
+                                    int &now_element,
+                                    std::vector<llvm::Value *> &now_ind) {
+  if (auto primitive_type = now_type->cast<PrimitiveType>()) {
+    TI_ASSERT((Type *)elements[now_element]->ret_type == now_type);
+    auto *gep = builder->CreateGEP(buffer_type, buffer, now_ind);
+    builder->CreateStore(llvm_val[elements[now_element]], gep);
+    now_element++;
+  } else if (auto struct_type = now_type->cast<StructType>()) {
+    int i = 0;
+    for (const auto &element_type : struct_type->elements()) {
+      now_ind.push_back(tlctx->get_constant(i++));
+      create_return(buffer, buffer_type, elements, element_type, now_element,
+                    now_ind);
+      now_ind.pop_back();
+    }
+  } else {
+    auto tensor_type = now_type->as<TensorType>();
+    int num_elements = tensor_type->get_num_elements();
```

Comment:
assert `tensor_type != nullptr` before use?

Reply:
The `as` method already does that for us.